### PR TITLE
feat(notifications): new tables for notification settings

### PR DESCRIPTION
**Pull request by scefali**
This PR adds two notification tables that are explained [here](https://www.notion.so/sentry/Fix-Notification-Setting-Data-Model-c510e815759f4c3f81cf4f62294235ca) in depth. Basically, we are splitting apart the existing `NotificationSetting` table into two tables: one to track the provider and another to track if notifications are enabled/disabled. The purpose is so we can remove the complicated mapping logic between the data model and the UI layer: each UI option will map to just a single row instead of one row per provider.

### CODE REVIEW

File: src/sentry/migrations/0529_new_notification_tables.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            name="NotificationSettingProvider",
+            fields=[
+                (
+                    "id",
+                    sentry.db.models.fields.bounded.BoundedBigAutoField(
+                        primary_key=True, serialize=False
+                    ),
+                ),
+                ("scope_type", sentry.db.models.fields.bounded.BoundedPositiveIntegerField()),
+                ("scope_identifier", sentry.db.models.fields.bounded.BoundedBigIntegerField()),
```
Comment (markstory) [inline src/sentry/migrations/0529_new_notification_tables.py:41]: If you can, using string enums will make adhoc querying and debugging easier as you don't have to keep the context of what enum int value maps to which type in your head.

File: src/sentry/models/__init__.py [hunk 2]
```diff
@@ -62,7 +62,9 @@
 from .latestappconnectbuildscheck import *  # NOQA
 from .latestreporeleaseenvironment import *  # NOQA
 from .lostpasswordhash import *  # NOQA
-from .notificationsetting import *  # NOQA
+from .notificationsetting import *  # NOQANOQA
```
Comment (markstory) [inline src/sentry/models/__init__.py:65]:
```suggestion
from .notificationsetting import *  # NOQA
```

File: src/sentry/models/notificationsettingbase.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            assert not (
+                self.user_id is None and self.team_id is None
+            ), "Notification setting missing user & team"
+        except AssertionError as err:
+            sentry_sdk.capture_exception(err)
+        super().save(*args, **kwargs)
+
+
+# REQUIRED for migrations to run
+from sentry.trash import *  # NOQA
```
Comment (markstory) [inline src/sentry/models/notificationsettingbase.py:91]: Is this true? It looks like carryover from the old notification settings models.

File: src/sentry/models/notificationsettingoption.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            assert not (
+                self.user_id is None and self.team_id is None
+            ), "Notification setting missing user & team"
+        except AssertionError as err:
+            sentry_sdk.capture_exception(err)
+        super().save(*args, **kwargs)
+
+
+# REQUIRED for migrations to run
+from sentry.trash import *  # NOQA
```
Comment (markstory) [inline src/sentry/models/notificationsettingoption.py:63]: Is this still required for the new models?

File: src/sentry/models/notificationsettingbase.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        ),
+        null=False,
+    )
+    value = BoundedPositiveIntegerField(
+        choices=(
+            (NotificationSettingOptionValues.NEVER, "never"),
+            (NotificationSettingOptionValues.ALWAYS, "always"),
+        ),
+        null=False,
+    )
```
Comment (markstory) [inline src/sentry/models/notificationsettingbase.py:75]: Both subclasses redefine the value column, does it need to be here was well?

File: src/sentry/models/notificationsettingbase.py (hunk 3 above)
Comment (scefali) [reply src/sentry/models/notificationsettingbase.py:91]: @markstory yea maybe not needed

File: src/sentry/models/notificationsettingbase.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        choices=(
+            (NotificationScopeType.USER, "user"),
+            (NotificationScopeType.ORGANIZATION, "organization"),
+            (NotificationScopeType.PROJECT, "project"),
+            (NotificationScopeType.TEAM, "team"),
+        ),
+        null=False,
+    )
+    scope_identifier = BoundedBigIntegerField(null=False)
+    team_id = HybridCloudForeignKey("sentry.Team", null=True, db_index=True, on_delete="CASCADE")
```
Comment (scttcper) [inline src/sentry/models/notificationsettingbase.py:47]: what are we doing with team_id?

File: src/sentry/migrations/0529_new_notification_tables.py (hunk 1 above)
Comment (scefali) [reply src/sentry/migrations/0529_new_notification_tables.py]: @markstory fixed to use strings

File: src/sentry/models/notificationsettingbase.py (hunk 6 above)
Comment (scefali) [reply src/sentry/models/notificationsettingbase.py:14]: @scttcper if the setting points to a team then we set the `team_id`

**Review by markstory — APPROVED** [review]
Schema looks good to me.

**Pull request merged**